### PR TITLE
CSHARP-4048: "Take" doesn't work with GroupBy.

### PR DESCRIPTION
**Pull request by rstam**

### CODE REVIEW

File: src/MongoDB.Driver/Linq/Linq3Implementation/Translators/ExpressionToAggregationExpressionTranslators/MethodTranslators/TakeMethodToAggregationExpressionTranslator.cs [hunk 1]
```diff
@@ -39,7 +39,7 @@ public static AggregationExpression Translate(TranslationContext context, Method
                     skipExpression = sourceSkipExpression.Arguments[1];
                 }
 
-                var sourceTranslation = ExpressionToAggregationExpressionTranslator.Translate(context, sourceExpression);
+                var sourceTranslation = ExpressionToAggregationExpressionTranslator.TranslateEnumerable(context, sourceExpression);
```
Comment (rstam) [inline src/MongoDB.Driver/Linq/Linq3Implementation/Translators/ExpressionToAggregationExpressionTranslators/MethodTranslators/TakeMethodToAggregationExpressionTranslator.cs:42]: Separate from this ticket I wonder if we should try to unify `Translate` and `TranslateEnumerable`.

It's just too easy to call `Translate` when we should call `TranslateEnumerable`

File: tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationTests/Jira/CSharp4048Tests.cs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+            {
+                "{ $group : { _id : '$_id', _elements : { $push: '$$ROOT' } } }",
+                "{ $project : { Id : '$_id', Take  : { $slice : ['$_elements', 1] }, _id : 0 } }"
+            };
+            AssertStages(stages, expectedStages);
+
+            var results = queryable.ToList();
+            results.Count.Should().Be(2);
+            results[0].ToJson().Should().Be("{ \"_id\" : 1, \"Take\" : [{ \"_id\" : 1, \"X\" : 11 }] }");
+            results[1].ToJson().Should().Be("{ \"_id\" : 2, \"Take\" : [{ \"_id\" : 2, \"X\" : 22 }] }");
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationTests/Jira/CSharp4048Tests.cs:52]: There must be a better way to compare two objects when the class doesn't override `Equals`...

**Comment by rstam** [comment]
`Take` might just be the tip of the iceberg.

I'm going to verify that *all* `IEnumerable` methods work with `IGrouping`.

**Comment by rstam** [comment]
The very first `IEnumerable` method I checked (`Aggregate`) has the same issue. So probably most others do also.

File: tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationTests/Jira/CSharp4048Tests.cs [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+            var documents = new[]
+            {
+                new C { Id = 1, X = 11 },
+                new C { Id = 2, X = 22 }
+            };
+            CreateCollection(collection, documents);
+
+            var queryable = collection.AsQueryable()
+                .GroupBy(x => x.Id)
+                .Select(x => new { Id = x.Key, Take = x.Take(1) });
```
Comment (JamesKovacs) [inline tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationTests/Jira/CSharp4048Tests.cs:39]: We need to add `.OrderBy(x => x.Id);` as tests are failing about 25% of the time due to documents coming back in the opposite order.

File: tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationTests/Jira/CSharp4048Tests.cs [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+            var stages = Translate(collection, queryable);
+            var expectedStages = new[]
+            {
+                "{ $group : { _id : '$_id', _elements : { $push: '$$ROOT' } } }",
+                "{ $project : { Id : '$_id', Take  : { $slice : ['$_elements', 1] }, _id : 0 } }"
+            };
```
Comment (JamesKovacs) [inline tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationTests/Jira/CSharp4048Tests.cs:46]: Add `"{ $sort : { Id : 1 } }"` to match the added `OrderBy` clause.

**Review by JamesKovacs — CHANGES_REQUESTED** [review]
See requested changes.

**Review by JamesKovacs — APPROVED** [review]
LGTM.

File: tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationTests/Jira/CSharp4048Tests.cs (hunk 4 above)
Comment (rstam) [reply tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationTests/Jira/CSharp4048Tests.cs:2207]: Done

**Pull request closed without merging**